components/pluralrules/src/operands.rs

```diff
@@ -0,0 +1,186 @@
+pub struct PluralOperands {
+    /// Absolute value of input
+    pub n: f64,
+    /// Integer value of input
+    pub i: u64,
+    /// Number of visible fraction digits with trailing zeros
+    pub v: usize,
+    /// Number of visible fraction digits without trailing zeros
+    pub w: usize,
+    /// Visible fraction digits with trailing zeros
+    pub f: u64,
+    /// Visible fraction digits without trailing zeros
+    pub t: u64,
+}
+
+impl<'a> TryFrom<&'a str> for PluralOperands {
+    type Error = &'static str;
+
+    fn try_from(input: &'a str) -> Result<Self, Self::Error> {
+        let abs_str = if input.starts_with('-') {
+            &input[1..]
+        } else {
+            &input
+        };
+
+        let absolute_value = f64::from_str(&abs_str).map_err(|_| "Incorrect number passed!")?;
+
+        let integer_digits;
+        let num_fraction_digits0;
+        let num_fraction_digits;
+        let fraction_digits0;
+        let fraction_digits;
+
+        if let Some(dec_pos) = abs_str.find('.') {
+            let int_str = &abs_str[..dec_pos];
+            let dec_str = &abs_str[(dec_pos + 1)..];
+
+            integer_digits =
+                u64::from_str(&int_str).map_err(|_| "Could not convert string to integer!")?;
+
+            let backtrace = dec_str.trim_end_matches('0');
+
+            num_fraction_digits0 = dec_str.len() as usize;
+            num_fraction_digits = backtrace.len() as usize;
+            fraction_digits0 =
+                u64::from_str(dec_str).map_err(|_| "Could not convert string to integer!")?;
+            fraction_digits = u64::from_str(backtrace).unwrap_or(0);
+        } else {
+            integer_digits = absolute_value as u64;
+            num_fraction_digits0 = 0;
+            num_fraction_digits = 0;
+            fraction_digits0 = 0;
+            fraction_digits = 0;
+        }
+
+        Ok(PluralOperands {
+            n: absolute_value,
+            i: integer_digits,
+            v: num_fraction_digits0,
+            w: num_fraction_digits,
+            f: fraction_digits0,
+            t: fraction_digits,
+        })
+    }
+}
+
+macro_rules! impl_integer_type {
+    ($ty:ident) => {
+        impl From<$ty> for PluralOperands {
+            fn from(input: $ty) -> Self {
+                // XXXManishearth converting from u32 or u64 to isize may wrap
+                PluralOperands {
+                    n: input as f64,
+                    i: input as u64,
+                    v: 0,
+                    w: 0,
+                    f: 0,
+                    t: 0,
+                }
+            }
+        }
+    };
+    ($($ty:ident)+) => {
+        $(impl_integer_type!($ty);)+
+    };
+}
+
+macro_rules! impl_signed_integer_type {
+    ($ty:ident) => {
+        impl TryFrom<$ty> for PluralOperands {
+            type Error = &'static str;
+            fn try_from(input: $ty) -> Result<Self, Self::Error> {
+                // XXXManishearth converting from i64 to isize may wrap
```

Comment:
what is "output" time here?

Reply:
What I mean is, if the params kept the value of x in the widest type that would contain any of the values, we could avoid the need to make the wrapping decision here.  However, since the fields `n` and `i` are public, we have to make that deliberation here.

Reply:
What is `x`?
`PluralOperands` are only used in plural rules resolution. I struggle to see a case where we'd benefit from lazy resolving them later, because I don't see a use case where we'd construct a `PluralOperands` but not want to get it's `i` or `n`.
And if we do, we'll want it for many rules as we test against them. So having them eagerly resolved seems like a cleaned API which caves to nice further optimizations.

For example, I can imagine scenarios where an instance of `PluralOperands` gets stored and passed around for use in `PluralRules`, and proc_macros will allow us to construct the operands at build time and have them validated and calculated at runtime for free.